arviz/stats/stats.py

```diff
@@ -349,84 +351,92 @@ def hpd(ary, credible_interval=None, circular=False, multimodal=False, skipna=Fa
         if not 1 >= credible_interval > 0:
             raise ValueError("The value of credible_interval should be in the interval (0, 1]")
 
-    if ary.ndim > 1:
-        hpd_array = np.array(
-            [
-                hpd(
-                    row,
-                    credible_interval=credible_interval,
-                    circular=circular,
-                    multimodal=multimodal,
-                )
-                for row in ary.T
-            ]
-        )
-        return hpd_array
-
     if multimodal:
-        if skipna:
-            ary = ary[~np.isnan(ary)]
-
-        if ary.dtype.kind == "f":
-            density, lower, upper = _fast_kde(ary)
-            range_x = upper - lower
-            dx = range_x / len(density)
-            bins = np.linspace(lower, upper, len(density))
-        else:
-            bins = get_bins(ary)
-            _, density, _ = histogram(ary, bins=bins)
-            dx = np.diff(bins)[0]
+        return _hpd_multimodal(ary, credible_interval, skipna)
+
+    if isinstance(ary, np.ndarray):
+        ary = convert_to_dataset(ary)
+
+    kwargs.setdefault("input_core_dims", [["chain", "draw"]])
+    kwargs.setdefault("output_core_dims", [["hpd"]])
+    func_kwargs = {
+        "credible_interval": credible_interval,
+        "circular": circular,
+        "skipna": skipna,
+        "out_shape": (2,),
+    }
+    return _wrap_xarray_ufunc(_hpd, ary, func_kwargs=func_kwargs, **kwargs)
```

Comment:
does wrap xarray ufunc work on inference data objects? Is the input supposed to be a dataset? I think now that we are improving and rewriting, it should behave similarly to `ess`/`rhat`, accept both inference data/dataset or ndarray.

Reply:
> does wrap xarray ufunc work on inference data objects? 

No, currently I don't think it works. We need to modify the `wrap_xarray_ufunc`.

Reply:
We may get to apply ufunc on idata objects, but clearly not right now, and probably not directly in hpd/ess functions.